### PR TITLE
[CONSVC-1955][Part 1] fix: enable mypy checks on merino directory

### PR DESCRIPTION
**Pull request by ncloudioj**
Took a first stab at appeasing `mypy` against everything in the `merino` directory. Partially enabled mypy in linting (`make lint`) and pre-commit checking.

@taddes Apologize in advance if this conflicts with anything in your working branch. `mypy` is indeed not 100% accurate, had to manually ignore several cases as I believed those are definitely false alarms.

### CODE REVIEW

File: merino/providers/__init__.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ async def init_providers() -> None:
             case ProviderType.ADM:
                 providers["adm"] = AdmProvider(
                     backend=(
-                        LiveBackend()
+                        LiveBackend()  # type: ignore
```
Comment (ncloudioj) [inline merino/providers/__init__.py:44]: Looks like `mypy` doesn't recognice `LiveBackend` and `TestBackend` as the underlying `RemoteSettingsBackend` protocol, already checked that all methods and argument types matched perfectly. Not sure why 🤷‍♂️

File: merino/providers/adm.py [hunk 2]
```diff
@@ -120,7 +120,7 @@ async def initialize(self) -> None:
 
     def _should_fetch(self) -> bool:
         """Check if it should fetch data from Remote Settings."""
-        return (
+        return (  # type: ignore
```
Comment (ncloudioj) [inline merino/providers/adm.py:123]: Somehow, `mypy` believed that the return value was of type `Any`.

File: merino/util/user_agent_parsing.py [hunk 3]
```diff
@@ -29,7 +29,7 @@ def _parse_browser(user_agent: dict[str, Any]) -> str:
             return "Firefox" if major is None else f"Firefox({version.rstrip('.')})"
         case {"family": browser}:
```
Comment (ncloudioj) [inline merino/util/user_agent_parsing.py:30]: Alternatively, we could match against a `str` value via `case {"family": str(browser)}`. But that would come with a runtime type check, whereas `cast(str, browser)` won't as we know `browser` here is indeed a "str".

File: merino/util/user_agent_parsing.py [hunk 4]
```diff
@@ -54,7 +54,7 @@ def _parse_os_family(operating_system: dict[str, Any]) -> str:
             return "other"
 
 
-def _parse_form_factor(device: dict[str, Any], os_family: str) -> str:
+def _parse_form_factor(device: dict[str, Any], os_family: str) -> str:  # type: ignore
```
Comment (ncloudioj) [inline merino/util/user_agent_parsing.py:57]: Looks like another false alarm. The same pattern match above worked just fine.

File: merino/providers/__init__.py (hunk 1 above)
Comment (taddes) [reply merino/providers/__init__.py:44]: I had the same issue. Annoying. I think when I did a little research, `mypy` can struggle with custom types that are imported.

File: merino/util/user_agent_parsing.py (hunk 4 above)
Comment (taddes) [reply merino/util/user_agent_parsing.py:57]: :thumbsup:

**Comment by ncloudioj** [comment]
Close this to favor #69

**Pull request closed without merging**